### ISSUE TITLE
[Bug]: WebDAV interface returns different displayname property when requesting PROPFIND of directory vs PROPFIND of item

### ISSUE THREAD

**Issue by rcwschaller** [nextcloud/server#37541]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

WebDAV clients can request information about files and sub-folders in (at least) two different ways:
1. by requesting a list of all items in a folder 
`PROPFIND /remote.php/webdav/my_folder/`
(notice the tailing dash in the request)
returns a xml with one <d:response> section per item in the folder, each containing a <d:displayname> attribute
2. by requesting properties of each item in the folder seperatly
`PROPFIND /remote.php/webdav/my_folder/my_subfolder`
(notice NO tailing dash in the request)
returns a xml with only one <d:response> section for my_subfolder, containing a <d:displayname> attribute

Both xmls will have a <d:response> section for my_subfolder, however in certain circumstances their <d:displayname> attribute is not equal (see steps to reproduce)

I believe response 1 is the intended and correct response (as it matches up with the contents of the <d:href> attribute) and response 2 is a bug.

### Steps to reproduce

Use web interface:
1. as user_A create a folder my_folder_A
2. as user_A share my_folder_A with user_B
3. as user_B rename my_folder_A to my_folder_B
Now depending on which account you are using the same folder content is accessible by two different path. So far behaviour as expected.

Use a WebDAV client (e.g. Nemo on a Linux Mint machine) and observe the network requests:
4. with the credentials of user_B request the base directory
4.1. `PROPFIND /remote.php/webdav/` is sent 
4.2. receives the xml including <d:response> section for the shared folder with a <d:displayname> attribute of "my_folder_B"
This matches up with the tail of the <d:href> attribute and the way the folder name is displayed in the web interface.
4.3. "my_folder_B" is displayed in the Nemo GUI
4.4. Nemo immediately goes on requesting properties of each item in the base directory with an individual request for each one. E.g. `PROPFIND /remote.php/webdav/my_folder_B`
[N.B.: I don't see a particular benefit in this behaviour of Nemo as it seems to gain no information by this step, that are not already discovered with step 4.1, but technically there is nothing wrong with these requests]
4.5 receives the xml only containing the <d:response> section for the shared folder with a <d:displayname> attribute of "my_folder_A"
This does NOT match up with the tail of the <d:href> attribute and the way the folder name is displayed in the web interface.
4.6 Nemo changes the displayed name in the GUI from "my_folder_B" to "my_folder_A"

### Expected behavior

<d:displayname> attribute of step 4.2 and 4.5 should at least be equal, probably matching up the the <d:href> tail.

### Installation method

Community Manual installation with Archive

### Nextcloud Server version

25

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### List of activated Apps

```shell
- activity: 2.17.0
  - admin_audit: 1.15.0
  - bruteforcesettings: 2.5.0
  - calendar: 4.3.1
  - circles: 25.0.0
  - cloud_federation_api: 1.8.0
  - comments: 1.15.0
  - contacts: 5.2.0
  - contactsinteraction: 1.6.0
  - dashboard: 7.5.0
  - dav: 1.24.0
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - firstrunwizard: 2.14.0
  - gpxedit: 0.0.14
  - impersonate: 1.12.0
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - mail: 2.2.5
  - notifications: 2.13.1
  - oauth2: 1.13.0
  - password_policy: 1.15.0
  - photos: 2.0.1
  - previewgenerator: 5.2.1
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - recommendations: 1.4.0
  - related_resources: 1.0.4
  - richdocuments: 7.1.2
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - spreed: 15.0.5
  - support: 1.8.0
  - systemtags: 1.15.0
  - text: 3.6.0
  - theming: 2.0.1
  - twofactor_admin: 4.1.9
  - twofactor_backupcodes: 1.14.0
  - twofactor_nextcloud_notification: 3.6.0
  - twofactor_totp: 7.0.0
  - twofactor_webauthn: 1.1.2
  - updatenotification: 1.15.0
  - user_status: 1.5.0
  - viewer: 1.9.0
  - weather_status: 1.5.0
  - workflow_script: 1.10.0
  - workflowengine: 2.7.0
```

**Comment by rcwschaller** [comment]
For documentation, here are also two curl requests to reproduce the issue:

Query:
curl --basic --user 'user:apppasswd' -i -X PROPFIND https://cloud.example.com/remote.php/webdav/ --upload-file - -H "Depth: 1" <<end
<?xml version="1.0"?>
<a:propfind xmlns:a="DAV:">
<a:prop><a:displayname/></a:prop>
</a:propfind>
end

Correct response (shortened):
[...]
<d:response><d:href>/remote.php/webdav/my_folder_B/</d:href><d:propstat><d:prop><d:displayname>my_folder_B</d:displayname></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response>
[...]

Query:
curl --basic --user 'user:apppasswd' -i -X PROPFIND https://cloud.example.com/remote.php/webdav/my_folder_B --upload-file - -H "Depth: 1" <<end
<?xml version="1.0"?>
<a:propfind xmlns:a="DAV:">
<a:prop><a:displayname/></a:prop>
</a:propfind>
end`

Wrong response (shortened):
[...]
<d:response><d:href>/remote.php/webdav/my_folder_B/</d:href><d:propstat><d:prop><d:displayname>my_folder_A</d:displayname></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response>
[...]

**Comment by rcwschaller** [comment]
Well, after digging around for sometime, here is what I found:

- The wrong response is delivered by `apps/dav/lib/Connector/Sabre/FilesPlugin.php` eventually. Function _handleGetProperties_ contains a comment for returning _self::DISPLAYNAME_PROPERTYNAME_ that it is written in a certain way to avoid performance issues. The comment does not state which other way would be the proper or even just other possible way. Hence I have no clue if this is completely unrelated to this issue or a performance optimisation, that forgot about this not-very-common(?) usecase and caused this issue.
- Assuming the performance issue to be unrelated, and the `$node->getName()` implementation is also applicable to this usecase, I tracked the wrong data across `apps/dav/lib/Connector/Sabre/Node.php`  `apps/dav/lib/Connector/Sabre/Directory.php` `apps/dav/lib/Connector/Sabre/ObjectTree.php` and `lib/private/Files/View.php` where it eventually in function _getCacheEntry_ comes out of a cache - and this is where I gave up as I did not want to dig into the whole caching system as well.
- For the moment I fixed it with a workaround in `apps/dav/lib/Connector/Sabre/FilesPlugin.php` by replacing the `$node->getName()` implementation with a `basename($node->getPath())`, patch file attached. **Obviously this does not fix the issue of why the wrong data gets delivered to the WebDAV system and instead only replaces it with the correct one from another source**

[patch.TXT](https://github.com/nextcloud/server/files/11288115/patch.TXT)

Hope this information helps whoever wants to find the root cause of this wrong data ;-)

**Comment by susnux** [comment]
> as it matches up with the contents of the <d:href> attribute

This is not necessary true.
For example groupfolders are named by ID e.g. `1` or `100`, but their displayname is something different (e.g. `Group 1 folder`).

**Comment by rcwschaller** [comment]
@susnux Thanks for the following up on such an old bug report.
I do believe that your comment and the fix #46694 do not actually relate to the originally reported problem here, but while retesting I discovered that the unexpected behaviour of this issue vanished somewhere between nextcloud 25 and 28.0.5 (so actually long before #46694).
So "closed" it is indeed :smile:
